#### Let's install faker library

In [ ]:
!pip install faker

#### import all the required packages

In [ ]:
from faker import Faker
from collections import namedtuple
import datetime
import random, time
fake = Faker()

### For question no 1 first let's create our namedtuple

In [475]:
profile = namedtuple('profile',fake.profile().keys())

In [476]:
def create_fake_profile(n):
    profile = namedtuple('profile',fake.profile().keys())
    return tuple([profile(**fake.profile()) for i in range(n)])


#### I have tried to create namedtuple of namedtuples but I saw that namedtuple can take only 255 arguments thats why I proceeded with creating tuple of namedtuples

In [483]:
fake_profile = create_fake_profile(10000)

#### After creating our 10000 fake profiles let's create a frequence of blood group[ dictionary and from there we can get max frequency of which blood group

In [477]:
def largest_blood_type(f_profile):
  blood_freq = dict()
  blood_grp_list = [profile.blood_group for profile in f_profile]
  {w: 1 if w not in blood_freq and not blood_freq.update({w: 1}) else blood_freq[w] + 1 if not blood_freq.update({w: blood_freq[w] + 1}) else 1 for w in blood_grp_list}
  return blood_freq, list(blood_freq.keys())[list(blood_freq.values()).index(max(blood_freq.values()))]
blood_freq, max_blood_grp = largest_blood_type(fake_profile)
print(blood_freq)
print(max_blood_grp)

{'A+': 1254, 'O-': 1266, 'O+': 1225, 'AB+': 1241, 'B+': 1218, 'AB-': 1238, 'B-': 1261, 'A-': 1297}
A-


#### Now from location of all profiles we will calculate mean location of all location

In [478]:
def mean_loc(f_profile, n):
  return sum([profile.current_location[0] for profile in f_profile])/n,sum([profile.current_location[1] for profile in f_profile])/n
loc = mean_loc(fake_profile, 10000)
print(loc)

(Decimal('-0.6416528696'), Decimal('-0.1882050778'))


#### Finally we are going to calculate amx age and avg age of all profiles

In [481]:
all_age = [(datetime.datetime.today().date() - profile.birthdate).days for profile in fake_profile]
print(max(all_age))
print(sum(all_age)/10000)

42366
21312.6039


### for question no 2 we are going to implement the above same exercize in list of dictionary.

In [466]:
def dict_profile(n):

  f_profile = [fake.profile() for i in range(n)]
  
  blood_freq = dict()
  blood_grp_list = [profile['blood_group'] for profile in f_profile]
  {w: 1 if w not in blood_freq and not blood_freq.update({w: 1}) else blood_freq[w] + 1 if not blood_freq.update({w: blood_freq[w] + 1})\
   else 1 for w in blood_grp_list}
  max_blood_grp_count = list(blood_freq.keys())[list(blood_freq.values()).index(max(blood_freq.values()))]

  mean_loc = sum([profile['current_location'][0] for profile in f_profile])/n,sum([profile['current_location'][1] for profile in f_profile])/n

  all_age = [(datetime.datetime.today().date() - profile['birthdate']).days for profile in f_profile]

  max_age = max(all_age)

  avg_age = sum(all_age)/n

  return max_blood_grp_count, mean_loc, max_age, avg_age

In [472]:
start2 = time.perf_counter()
for i in range(2):
  _,_,_,_ = dict_profile(10000)
end2 = time.perf_counter()
delta2 = end2-start2
print(delta2)

38.2584211689973


#### After implementing in dictionary we also checked the time taked by it and we saw that dictionary is taking around 38.2584 sec

In [463]:
def namedtuple_profile(n):

  profile = namedtuple('profile',fake.profile().keys())
  f_profile = tuple([profile(**fake.profile()) for i in range(n)])

  blood_freq = dict()
  blood_grp_list = [profile.blood_group for profile in f_profile]
  {w: 1 if w not in blood_freq and not blood_freq.update({w: 1}) else blood_freq[w] + 1 if not blood_freq.update({w: blood_freq[w] + 1})\
   else 1 for w in blood_grp_list}
  max_blood_grp_count = list(blood_freq.keys())[list(blood_freq.values()).index(max(blood_freq.values()))]

  mean_loc = sum([profile.current_location[0] for profile in f_profile])/n,sum([profile.current_location[1] for profile in f_profile])/n

  all_age = [(datetime.datetime.today().date() - profile.birthdate).days for profile in f_profile]

  max_age = max([(datetime.datetime.today().date() - profile.birthdate).days for profile in f_profile])

  avg_age = sum(all_age)/n

  return max_blood_grp_count, mean_loc, max_age, avg_age

In [473]:
start1 = time.perf_counter()
for i in range(2):
  _,_,_,_ = namedtuple_profile(10000)
end1 = time.perf_counter()
delta1 = end1-start1
print(delta1)

38.25425250000262


In [485]:
delta2-delta1>0

True

#### After checking the same with named tuple we saw that named tuple is performing little bit better as it is taking around 38.2542 sec. So, we can say that namedtuple is performing little bit better than dictionary

#### For the 3rd assignment we have created namedtuples of namedtuple of 100 fake companies. Here below we have only used named tuples.

In [405]:
company = namedtuple('company','company_name symbol open high close')

def create_company_tuple():

    company_name = fake.company()

    symbol = company_name[0:4].upper()

    open = random.randint(10,4000)

    high = int(open * random.uniform(.9,1.1))

    close = int(high * random.uniform(0.9,1))

    return company(company_name,symbol,open,high,close)

In [406]:
company_tuple = namedtuple('company_tuple',list(range(100)),rename=True)

#### Since we have to assign some weight to them for that we have created a named tuple of normalized weights and assigned each company one weight

In [408]:
weight_list = [round(random.uniform(0,1),2) for i in range(100)]
weight = namedtuple('weight',list(range(100)),rename=True)
normalized_weight = weight._make([i/sum(weight_list)for i in weight_list])
print(normalized_weight)

weight(_0=0.014925373134328358, _1=0.003383084577114428, _2=0.0029850746268656717, _3=0.015522388059701494, _4=0.0053731343283582094, _5=0.010945273631840797, _6=0.009353233830845771, _7=0.005970149253731343, _8=0.01671641791044776, _9=0.011144278606965175, _10=0.004776119402985075, _11=0.011144278606965175, _12=0.011343283582089551, _13=0.0021890547263681594, _14=0.01691542288557214, _15=0.014129353233830844, _16=0.007562189054726368, _17=0.009154228855721393, _18=0.006965174129353233, _19=0.012338308457711443, _20=0.005970149253731343, _21=0.003383084577114428, _22=0.012736318407960199, _23=0.0191044776119403, _24=0.017114427860696516, _25=0.00955223880597015, _26=0.01592039800995025, _27=0.0025870646766169153, _28=0.010149253731343283, _29=0.01671641791044776, _30=0.004378109452736319, _31=0.011144278606965175, _32=0.008557213930348258, _33=0.015323383084577116, _34=0.011542288557213929, _35=0.018706467661691543, _36=0.01592039800995025, _37=0.01572139303482587, _38=0.01671641791044

#### With the above mentioned weights we are going to create one more modified named tuple and create our final company list with normalized weight

In [411]:
stock = namedtuple('stock',company._fields + ('stock_weight',))
ct = company_tuple._make([stock(*create_company_tuple(),normalized_weight[i]) for i in range(100)])
print(ct)

company_tuple(_0=stock(company_name='Morris Group', symbol='MORR', open=816, high=790, close=781, stock_weight=0.014925373134328358), _1=stock(company_name='Stone, Thomas and Estes', symbol='STON', open=2951, high=2715, close=2698, stock_weight=0.003383084577114428), _2=stock(company_name='Hudson, Lee and Reese', symbol='HUDS', open=2293, high=2223, close=2185, stock_weight=0.0029850746268656717), _3=stock(company_name='Miller, Saunders and Morrow', symbol='MILL', open=1688, high=1770, close=1712, stock_weight=0.015522388059701494), _4=stock(company_name='Landry-Oconnor', symbol='LAND', open=544, high=592, close=566, stock_weight=0.0053731343283582094), _5=stock(company_name='Stewart, Peters and Daniels', symbol='STEW', open=2708, high=2614, close=2500, stock_weight=0.010945273631840797), _6=stock(company_name='Craig-Cole', symbol='CRAI', open=787, high=765, close=703, stock_weight=0.009353233830845771), _7=stock(company_name='Scott-Newman', symbol='SCOT', open=2858, high=3135, close=3

#### FInally we are going to check market open, high and close price overall below and store it in a named tuple:

In [412]:
stock_market_details = namedtuple('stock_market_details','open high close')

market_open, market_high, market_close = 0,0,0
for i in range(len(ct)):
  market_open += (ct[i].open * ct[i].stock_weight)
  market_high += (ct[i].high * ct[i].stock_weight)
  market_close += (ct[i].close * ct[i].stock_weight)

stock_market = stock_market_details(open = market_open, high = market_high, close = market_close) 
print(stock_market)

stock_market_details(open=2100.7781094527363, high=2093.2278606965174, close=2002.883383084578)


#### In the above we can see our final namedtuple with market, open high and close price